lib/org/chromium/webidl/Diff.js

```diff
@@ -0,0 +1,247 @@
+        if (['boolean', 'number', 'string'].includes(leftType)) {
+          if (left !== right)
+            this.addChunk_(id, this.DiffStatus.VALUES_DIFFER, left,
+                right, opt_changes);
+          return;
+        }
+
+        // Left and right are either Array or Objects at this point.
+        if (isLeftArray) {
+          this.arrayCmp_(left, right, id, opt_changes);
+        } else {
+          foam.assert(left.cls_ && right.cls_,
+              'Diff: Cannot find class property of object.');
+
+          var leftProps = left.cls_.getAxiomsByClass(foam.core.Property);
+          var rightProps = right.cls_.getAxiomsByClass(foam.core.Property);
+          
+          // Determine if left and right have the same properties.
+          if (foam.util.compare(leftProps, rightProps) !== 0) {
+            // Cannot perform a deep comparison of the objects.
+            this.addChunk_(id, this.DiffStatus.VALUES_DIFFER, left,
+                right, opt_changes);
+            return;
+          }
+
+          // leftProp and rightProp are same at this point.
+          // Perform diff on each of the objects properties.
+          leftProps.forEach(function(prop) {
+            var propName = prop.name;
+            if (this.excludedProps.includes(propName)) return;
+
+            var nextId = Object.assign({}, id,
+                { path: id.path.concat(`.${propName}`) });
+            this.diff_(left[prop], right[prop], nextId, opt_changes);
+          }.bind(this));
+        }
+      },
+    },
+    {
+      documentation: `Performs a comparison between members of left and
+        members of right. If no differences are found between two objects,
+        they are considered to be matched and removed from the remaining
+        comparisons. Once all members have been compared, a DiffChunk is
+        generated for all unmatched members.`,
+      name: 'arrayCmp_',
+      returns: '',
+      code: function arrayCmp_(left, right, id, opt_changes) {
+        // Items will be removed from the array once matched.
+        // Work on clones to prevent altering original data.
+        left = foam.Array.clone(left);
+        right = foam.Array.clone(right);
+
+        var matched = false;
+        for (var i = 0; i < left.length; i++) {
+          for (var j = 0; j < right.length; j++) {
+            // Get id field of left and right.
+            var leftName = left[i].id;
+            var rightName = right[j].id;
+
+            // If names differ, they are definitely not the same.
+            if (leftName !== rightName) continue;
+            if (!leftName && !rightName) {
+              // These items have no names. Perform a diff with changes
+              // to determine whether these items are the same.
+              var changes = opt_changes || [];
+              var initLen = changes.length;
+
+              var nextId = Object.assign({}, id,
+                  { path: id.path.concat(`[${leftName}]`) });
```

Comment:
Extended attributes are always UpperCamelCase where members are almost always lowerCamelCase, but the latter is convention. This is syntactically legit:
```WebIDL
partial interface Document {
 [Unforgeable] readonly attribute DOMString Unforgeable;
};
```

I suspect there are zero cases like this in the wild, though.